### PR TITLE
Backfill script to update content_area, curriculum_umbrella and topic tags fields for existing curriculum.

### PR DESCRIPTION
**Pull request by vijayamanohararaj**
As part of requirements captured at https://docs.google.com/document/d/1u4kFMpiktoTv7reTBXgWjW0MXbKH3V3stO9s7PWS4ss/edit#heading=h.qi8xxyq72bct
 two new fields were introduced (in addition to the existing curriculum_umbrella field) to bring about more granular categorization of curriculum units. This change adds a script that would backfill existing units with values for these new fields and some updates to the existing curriculum umbrella field based on a csv input provided by curriculum team.

## Links

JIRA: https://codedotorg.atlassian.net/browse/TEACH-1365 and https://codedotorg.atlassian.net/browse/TEACH-1367

## Testing story

- Validated these changes locally in both whatif mode and actual execution
- Identified a few scripts that would fail due to two issues
 1. The script doesn't exist (spot checked a few and confirmed they don't exist in level builder as well)
 2. The script has a name (with uppercase letter) that fails validation rules

## Deployment strategy

- Wait for https://github.com/code-dot-org/code-dot-org/pull/62397 to get deployed before merging

## Follow-up work

- Run this on levelbuilder, well after work hours, will additionally isolate these changes into a single commit to make it easy to revert if needed

## Privacy

N/A

## Security

N/A

## Caching

N/A

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+# Take link to CSV file as input
+def backfill_script_data_categories(file_path)
+  raise unless Rails.application.config.levelbuilder_mode
+
+  puts "Reading csv from #{file_path}"
+  scripts_to_update = CSV.foreach(file_path, headers: true).map do |row|
+    [row["Script name"],
+     {
+       new_initiative: row["Initative"],
```
Comment (vijayamanohararaj) [inline bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb]: Make these into constants

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (!initiative_val.empty? && !$valid_curriculum_umbrella.include?(initiative_val)) ||
+        (!content_area_val.empty? && !$valid_content_area.include?(content_area_val)) ||
+        (!topic_tags_val.empty? && !$valid_topic_tags.include?(topic_tags_val))
+      puts "Updates for #{script_name} does not include a valid value for one of the three fields."
+      next
+    end
+
+    next if whatif_mode
+
+    script.curriculum_umbrella = new_values[:new_initiative] unless initiative_val.empty?
```
Comment (vijayamanohararaj) [inline bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb]: reuse the variable

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    topic_tags_val = new_values[:topic_tags].to_s
+
+    if (!initiative_val.empty? && !$valid_curriculum_umbrella.include?(initiative_val)) ||
+        (!content_area_val.empty? && !$valid_content_area.include?(content_area_val)) ||
+        (!topic_tags_val.empty? && !$valid_topic_tags.include?(topic_tags_val))
+      puts "Updates for #{script_name} does not include a valid value for one of the three fields."
+      next
+    end
+
+    next if whatif_mode
```
Comment (vijayamanohararaj) [inline bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb]: Add some logging for whatif mode

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    script.write_script_json
+  end
+end
+
+if ARGV.empty? || ARGV.length >2
+  warn "Usage: backfill_script_curriculum_classification_data.rb <path_to_csv_with_data> -whatif"
+  return
+end
+
+backfill_script_data_categories(ARGV[0], ARGV.length == 2 && ARGV[1] == "-whatif")
```
Comment (etaderhold) [inline bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb]: I think for typo-safety it's common to make the "what if" mode be the default, and the mode to actually commit changes requires that you supply a special properly-spelled flag. You have the opposite here.

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    end
+
+    if whatif_mode
+      puts "Updates for #{script_name} = [#{initiative_val}] [#{content_area_val}] [#{topic_tags_val}]"
+      next
+    end
+
+    script.curriculum_umbrella = initiative_val unless initiative_val.nil?
+    script.content_area = content_area_val unless content_area_val.nil?
+    script.topic_tags = topic_tags_val unless topic_tags_val.nil?
```
Comment (etaderhold) [inline bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb:56]: nit: you already did all these nil checks above.

**Review by etaderhold — APPROVED** [review]

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb (hunk 4 above)
Comment (vijayamanohararaj) [reply bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb]: Thanks, yes that makes sense. Updated to have "what if" as the default.

File: bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb (hunk 5 above)
Comment (vijayamanohararaj) [reply bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb:56]: The above check should only block execution if the value isn't nil and is not part of a valid list of values. So, we could still have nil values here. Pl let me know if you see a reason why that wouldn't be the case.

Comment (etaderhold) [reply bin/oneoff/backfill_data/backfill_script_curriculum_classification_data.rb:56]: Ah, good point. You're right.

**Review by etaderhold — APPROVED** [review]

**Pull request merged**